Add new flat AUI tab art

vadz opened the pull request:
As discussed in [this post](https://groups.google.com/g/wx-dev/c/mLeHQZ69ELs) the current situation with AUI art providers is not optimal. I've added a new art provider which looks a bit better (although I admit that it doesn't look that great, in spite of all my efforts), while being simpler (which hopefully means that maybe others will be able to improve it).

Here some screenshots of it under Linux in light ![image](https://github.com/user-attachments/assets/9472a86f-efb9-4b41-b61f-71fbcb45e48b)
and dark
![image](https://github.com/user-attachments/assets/59924544-71a9-4571-a7b8-b16716399fa1)
modes and also under Windows, where I also show it in multiline mode:
![image](https://github.com/user-attachments/assets/09b4982d-c97d-42ff-9f3b-bd0d59d3c4e7)
![image](https://github.com/user-attachments/assets/22c2cafd-5413-408d-897d-69e065bdec3f)

I'd like to make this the default provider for 3.3.0, any thoughts?

Note that this includes #25274 which I had submitted as a separate PR as it's not logically part of this one, even if I did while working on this.

vadz committed:
Factor out common tab art classes parts into a common base class

Extract code duplicated between wxAuiGenericTabArt and wxAuiSimpleTabArt
into a new wxAuiTabArtBase class which could (and will) be used as a
base class for other wxAuiTabArt implementations in the future.

No real changes yet.

vadz committed:
Refactor code for finding contrasting colour

Leave just a single wxAuiEnsureSufficientContrast() function instead of
two different functions that always had to be used in tandem and
resulted in less clear code.

No real changes.

vadz committed:
Move ShowDropDown() implementation to wxAuiTabArtBase too

This function was implemented slightly differently in the existing tab
art subclasses, with wxAuiSimpleTabArt using check menu items and
positioning the menu differently, but also not accounting for the
ampersands in the page caption.

None of these differences looked intentional, and having many check
items in the menu didn't look good, so reuse wxAuiGenericTabArt code for
both classes.

This commit is best viewed with Git --color-moved option.

vadz committed:
Simplify ShowDropDown() code and remove wxAuiCommandCapture

This code had been written before GetPopupMenuSelectionFromUser() was
available, but now that it is, the same thing can be done in much
simpler way.

No real changes.

vadz committed:
Prettify ShowDropDown() code a little

Use symbolic constant instead of hardcoded 1000.

Use range-for loop.

No real changes.

vadz committed:
Also move border-related functions to wxAuiTabArtBase

Don't duplicate them between wxAuiGenericTabArt and wxAuiSimpleTabArt,
they are the same in both classes and the default implementation is
reasonable enough to be reused in new classes too.

vadz committed:
Remove unused variable from wxAuiGenericTabArt

This wxBitmapBundle was never used, so just remove it.

vadz committed:
Avoid repeated conversion from wxBitmap to wxBitmapBundle

Micro-optimization in wxAuiGenericTabArt::GetBestTabCtrlSize(): instead
of converting a wxBitmap to wxBitmapBundle inside the loop, potentially
many times, do it just once outside of it.

No real changes.

vadz committed:
Share DrawButton() implementation between tab art classes too

The existing implementations were slightly different, but the difference
can be extracted into a new DrawButtonBitmap() virtual function, which
is trivial by default and whose reimplementation in wxAuiSimpleTabArt
already existed as a global DrawButtons() function in the source, so it
just had to be renamed and moved.

The new version duplicates less code and will allow reusing DrawButton()
in other tab art classes.

vadz committed:
Move indent size into IndentPressedBitmap() itself

It seems to make sense to always indent the bitmap by the same amount,
so put wxSize(1, 1) into the function doing it instead of passing it to
it on each call site.

No real changes.

vadz committed:
Add wxAuiFlatTabArt art provider

This provider implements a non-glossy look, unlike wxAuiGenericTabArt,
but is still not as weird/ugly as wxAuiSimpleTabArt.

For now this provider has to be chosen explicitly, but the plan is to
make it the default one in the future.

vadz committed:
Create helper wxAuiDimColour() function

This was used for dimming the "Close" button in wxAuiGenericTabArt but
can, and will, be used for dimming other colours too.

No changes yet, this is a pure refactoring.

vadz committed:
Use hues of blue for highlight colours in MSW dark mode

Using grey colours meant that we didn't have almost any non-grey system
colours in the dark mode, unlike in the light one, in which all the
high/hot light colours were bluish.

Change these colours to (different) hues of blue in dark mode as well to
make them more appropriate for actually highlighting something.

vadz committed:
Improve wxAuiDefaultTabArt documentation

Document wxAuiGenericTabArt separately, in order to prepare for changing
wxAuiDefaultTabArt to be something different from it.

AliKet commented:
I tested this PR under Windows 10 + wxQt 6.8.2 (light & dark modes):

![aui-light](https://github.com/user-attachments/assets/5838de32-3995-4783-978b-d11c775b472f)

![aui-dark](https://github.com/user-attachments/assets/7e1d7e55-0921-4791-9e64-4e0d686428d2)

vadz commented:
Thanks!

Is there some colour that would work better for highlighting than `HOTLIGHT` in wxQt? Or maybe it should be changed there, as in wxMSW dark mode? Because black definitely doesn't look like a great choice for highlighting, especially in dark mode.

AliKet commented:
`wxSYS_COLOUR_HOTLIGHT` is not implemented under `wxQt`:

> wxSYS_COLOUR_HOTLIGHT: Colour for a hyperlink or hot-tracked item

and here is a patch that fixes it:

```C++
diff --git a/src/qt/settings.cpp b/src/qt/settings.cpp
index 04cd02d6bc..661e59a995 100644
--- a/src/qt/settings.cpp
+++ b/src/qt/settings.cpp
@@ -120,6 +120,9 @@ wxColour wxSystemSettingsNative::GetColour(wxSystemColour index)
             break;

         case wxSYS_COLOUR_HOTLIGHT:
+            color = pal.color(QPalette::Link);
+            break;
+
         case wxSYS_COLOUR_GRADIENTACTIVECAPTION:
         case wxSYS_COLOUR_GRADIENTINACTIVECAPTION:
             // TODO


```

I have updated the images above to reflect this change.

vadz commented:
Looks much nicer, thanks! I'll push the wxQt fix separately.

vadz committed:
Don't reserve space for hidden tab buttons in wxAuiFlatTabArt

This was initially done in order to show the hidden buttons on hover,
but as this feature hasn't been implemented yet, avoid wasting space on
the buttons that are never shown needlessly.

Blake-Madden commented:
Looks good under Windows 11 (wxMSW).
